app/models/concerns/hostext/taxonomies.rb

```diff
@@ -0,0 +1,20 @@
+module Hostext
+  module Taxonomies
+    extend ActiveSupport::Concern
+
+    included do
+      validates :organization_id, :presence => true, :if => ->(host) { host.managed? && SETTINGS[:organizations_enabled] }
+      validates :location_id,     :presence => true, :if => ->(host) { host.managed? && SETTINGS[:locations_enabled] }
+      after_validation :fix_hostgroup_mismatches, :if => ->(host) { host.hostgroup.present? }
+    end
+
+    def fix_hostgroup_mismatches
+      Taxonomy.enabled_taxonomies.each do |taxonomy|
+        host_taxonomy = public_send(taxonomy.singularize.to_sym)
+        next if host_taxonomy.blank? ||
+          hostgroup.public_send(:"#{taxonomy}").include?(host_taxonomy)
```

Comment:
nitpick: you can work on the join table direct instead of looping though hostgroup, organizations and locations.

e.g. to save some extra un-required sql you can
```ruby
TaxableTaxonomy.first_or_create(:taxonomy => taxonomy, :taxable => hostgroup)
```
or even 
```
TaxableTaxonomy.first_or_create(:taxonomy_id => host.taxonomy_id, :taxable_id => host.hostgroup_id, :taxable_type => 'Hostgroup')
```